[sjaqjwor(이승기)] step4 구간삭제 PR 요청드립니다.

sjaqjwor opened the pull request:
안녕하세요

마지막 step4 PR 요청 드립니다.

그동안 많은 피드백 주셔서 많이 배우고 , 생각하는 시간이 되었습니다.

이번 step도 많은 피드백 부탁드립니다.

감사합니다 😄

sjaqjwor committed:
docs (readme) : 구간 제거 기능에 대한 요구 사항 분석

sjaqjwor committed:
test : 구간에 존재하지 않는 역을 삭제하려면 삭제 할 수 없다.

sjaqjwor committed:
test : 구간에 존재하지 않는 역을 삭제하려면 삭제 할 수 없다.

sjaqjwor committed:
feat : 구간에 추가되지 않는 역은 삭제 할 수 없다.

sjaqjwor committed:
test : 구간이 하나인 경우에는 삭제 할 수 없다.

sjaqjwor committed:
feat : 구간이 하나 인 경우 삭제 할 수 없는 기능 추가

sjaqjwor committed:
test : 상행 종점을 삭제 할 수 있다.

sjaqjwor committed:
feat : 상행 종점을 삭제하는 기능 추가

sjaqjwor committed:
test : 하행 최종 종점을 제거 할 수 있다.

sjaqjwor committed:
test : 중간 section에 있는 구간을 제거 할 수 있다.

sjaqjwor committed:
feat : 중간 구간 삭제하는 기능

sjaqjwor committed:
refactor(피드백) : 메소드 명칭 변경

seokhongkim commented on src/test/java/nextstep/subway/line/LineSectionAcceptanceTest.java:
```diff
@@ -156,14 +156,110 @@ void addSectionMiddleFailedDistance(long distance) {
         assertThat(saveResponse.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
     }
 
+    @DisplayName("삭제하고자 하는 역이 구간에 존재 하지 않아서 삭제 할 수 없다.(기본 강남 - 성수역 구간만 존재)")
```
지하철 역을 삭제하는 인수 테스트를 케이스별로 꼼꼼하게 잘 작성해주셨네요 👍

sjaqjwor replied:
감사합니다 😄

seokhongkim commented on src/test/java/nextstep/subway/line/LineSectionAcceptanceTest.java:
```diff
@@ -156,14 +156,110 @@ void addSectionMiddleFailedDistance(long distance) {
+    }
+
+    @DisplayName("중간 구간을 삭제 할 수 있다. (강남 - 성수역 - 홍대구간- 구의 존재)")
+    @ParameterizedTest
```
여기는 테스트 케이스가 하나만 있으니 `@Test` 를 사용해보면 어떨까요?

sjaqjwor replied:
예상 결과를 비교하기 위해 @ParameterizedTest 사용했는데

테스트 메소드 내에서 해결 하도록 하겠습니다.

감사합니다 😄 

@ParameterizedTest는 여러 케이스가 있을 때 사용하도록 하겠습니다.

seokhongkim commented on src/main/java/nextstep/subway/exception/DeleteSection.java:
```diff
@@ -0,0 +1,7 @@
+package nextstep.subway.exception;
+
+public class DeleteSection extends RuntimeException {
```
예외 클래스는 이름만 보고 예외라는 것을 알 수 있도록 `Exception` suffix 를 추가해보셔도 좋겠습니다.

또한 `구간 삭제` 예외보다 `구간 삭제 실패` 예외가 조금 더 상황을 잘 설명해주는 것 같기도 합니다. :)

sjaqjwor replied:
Exception을 붙여주는게

더 확실하게 인지를 할 수 있겠네요

감사합니다. 😄

seokhongkim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -81,10 +71,28 @@ public boolean anyMatch(Section section) {
         return this.upStation.equals(section.getUpStation());
     }
 
+    public boolean inCludeStationInSection(Station deleteStation) {
```
`section.includeStationInSection()`의 형태로 호출을 할 것 같네요.
`section`을 대상으로 메세지를 보내는 것이니 `InSection`은 없어도 맥락상 `Section`이 `station`을 가지고 있는지 물어보는 메서드라고 이해할 수 있을 것 같은데, 승기님이 보셨을 때는 어떤가요? :)
또한 컬렉션 API에서 어떤 원소를 포함하고 있는지 확인하는 API의 이름으로 `contains`를 주로 사용하는데요. 비슷한 이름을 구간 API에서도 제공해주면 처음 사용하는 사람도 API를 찾거나, 이름을 보고 행위를 유추할 때 더 쉽게 생각할 수 있을 것 같아요!

sjaqjwor replied:
section을 대상으로 메세지를 보내는 것이니 InSection은 없어도 맥락상 Section이 station을 가지고 있는지 물어보는 메서드라고 이해할 수 있을 것 같은데, 승기님이 보셨을 때는 어떤가요? :)
- 메소드를 작성할떄 최대한 어떤건지 포함시키기 위해 inSection을 붙이게 되었습니다. 
- @seokhongkim 님이 말씀 하신대로 InSection은 불필요할꺼 같습니다. 결국 호출하는 애는 Section이 될테니 메소드 의미상으로도 불필요해 보입니다. 
- inClunde보다는 contains가 더 어울리는것 같습니다. 다른 API(List 등등)만 보아도 contains를 범용적으로 사용하고 있으니 해당 naming으로 따라가는게 좋을 꺼 같습니다.

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -164,4 +177,51 @@ private List<Section> createSectionRecursive(Section firstSection, Map<Station,
+    }
+
+    private void deleteDownStationTerminus(LinkedList<Section> sections, Station deleteStation) {
+        Section downStationTerminus = sections.getFirst();
```
하행 종점이니 `sections.getLast()`를 사용해야 할까요? :)

맞다면, 이 메서드가 정상적으로 동작하지 않았을 것 같은데요.
왜 인수 테스트에서 인지하지 못 했는지 고민해보신 후, 인수 테스트를 먼저 수정해서 실패하도록 하고 코드를 고쳐보시면 좋겠습니다.

sjaqjwor replied:
해당 메소드를 호출하는 deleteSection 메소드 부터 잘못된것 같습니다 😢 

해당 부분을 테스트코드를 하나씩 짜면서 구현하다보니

제대로 작성하지 못한 것 같습니다.

이부분 수정해서 다시 피드백 드리겠습니다.

피드백 주시는데 이런거 신경쓰지 못해서 죄송합니다 😢

sjaqjwor replied:
@seokhongkim 

테스트 -> 구현 -> 테스트 다시 실행 해당 순서로 하고 다음 기능을 구현할때는 다른 케이스들을 검사하지 않아서 

발생했습니다.

또한 그냥 테스트에서 인지 못한 이유는 검증 단계에서 더 명확한 검증을 했어야 했는데 단순 거리만

비교 하다보니 발생했습니다 😢

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -23,6 +23,22 @@ public void addSection(Line line, Station upStation, Station downStation, Long d
+        deleteUpStationTerminus(sections, deleteStation);
+        deleteDownStationTerminus(sections, deleteStation);
+        deleteMiddle(sections, deleteStation);
+
```
불필요한 빈 줄로 보이네요.

sjaqjwor replied:
제거하도록 하겠습니다 😢

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -164,4 +177,51 @@ private List<Section> createSectionRecursive(Section firstSection, Map<Station,
     }
+
+
+    private void deleteMiddle(LinkedList<Section> sections, Station deleteStation) {
```
상행 종점 또는 하행 종점인 경우 `upStationSection`, `downStationSection` 을 조회하는 로직에서 항상 예외가 발생할 것 같습니다.

sjaqjwor replied:
@seokhongkim 

원래 의도했던것이 아니였는데... 제대로 코드가 동작하지 못했네요

해당 부분도 add , delete에서 공통으로 사용할 수 있는 부분을 메소드로 추출하고

분기 처리해서 해당 이유 처리하도록 하겠습니다. 😢

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -23,6 +23,22 @@ public void addSection(Line line, Station upStation, Station downStation, Long d
+        validateSectionsDefaultSize();
+
+        LinkedList<Section> sections = new LinkedList<>(allSortedSections());
+        deleteUpStationTerminus(sections, deleteStation);
```
역 삭제 요청이 왔을 때 대상 역은 상행 종점, 하행 종점, 중간 역 중 하나의 케이스에만 해당할 것 같아요.
이 부분도 분기를 해보면 어떨까요? 그럼 역을 추가하는 곳과 비슷한 로직이니 메서드를 재사용해볼 수 있겠네요.

sjaqjwor replied:
@seokhongkim 

메소드 재사용으로 피드백 반영하도록 하겠습니다.

seokhongkim commented on src/test/java/nextstep/subway/line/LineSectionAcceptanceTest.java:
```diff
@@ -156,14 +156,110 @@ void addSectionMiddleFailedDistance(long distance) {
+
+    @DisplayName("상행 종점을 삭제 할 수 있다. (강남 - 성수역 - 홍대구간 존재)")
+    @Test
+    void deleteUpStationTerminus() {
```
상행 종점 삭제와 하행 종점 삭제 테스트는 실패하는 것으로 보이네요!

sjaqjwor replied:
@seokhongkim 

해당 부분 이슈도 다시 수정 하도록 하겠습니다.

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -23,6 +23,22 @@ public void addSection(Line line, Station upStation, Station downStation, Long d
+        return createSections(upStationTerminus);
+    }
+
+    public void deleteSection(Station deleteStation) {
```
메서드 안에서 사용하는 private 메서드의 위치가 너무 멀리 있어서, 코드를 볼 때 위아래로 스크롤 이동이 많이 필요하네요. 😢 
지난 PR에서 이야기 나눈 메서드 배치 순서를 고려하여 위치를 변경해보면 어떨까요?

sjaqjwor replied:
아... 이번에는 꼭 반영하도록 하겠습니다.

제가 봐도 헛 스크롤이 많이 가네요..ㅠㅠ

seokhongkim reviewed (changes requested):
승기님, 4단계 미션 잘 진행해주셨습니다. 👍 
코드를 보고 궁금한 점과 고민해보시면 좋을 것 같은 내용을 코멘트로 남겼습니다.
코멘트 확인 부탁드립니다. 🙏

sjaqjwor committed:
refactor (피드백) : Custom Excepion 명칭에 exception 추가

sjaqjwor committed:
refactor(피드백) : @ParmeterizedTest -> @Test 로 변경

sjaqjwor committed:
refactor(피드백) : include 보다는 범용적인 conatins를 명칭에 포함

sjaqjwor committed:
refactor(피드백) : add , delete에서 공통으로 사용 가능한 메소드로 추출 후 delete에서 분기처리

sjaqjwor committed:
refactor(피드백) : 종점 하행역 , 종점 상행역 , 중간 구역 삭제에 대한 이슈 처리(자세한 내용 커밋 참조)

update:

- deleteSection 분기 처리
- getFirst -> getLast로 변경
- 테스트코드 통과 이유
   - 단순 첫번째 결과에서 거리만 비교 했기 때문에 통과하게 됨
   - 정확한 section이 몇 개인지 확인하는 테스트 코드 추가

sjaqjwor committed:
refactor(test) : 테스트 코드 검증 추가

sjaqjwor committed:
refactor(피드백) : 메소드 순서 위치 변경

sjaqjwor committed:
refactor(test) : 검증 부분 리팩토링

sjaqjwor commented:
안녕하세요!

PR 날리기전에 최종 점검 하지 못한점

진심으로 죄송합니다..

피드백 반영했습니다

감사합니다 😢

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -12,9 +9,12 @@
+import static nextstep.subway.exception.ErrorStatus.SECTION_DEFAULT_SIZE;
+
 @Embeddable
 public class Sections {
```
`Section`, `Sections` 클래스에 도메인 로직이 많이 추가되었으니 단위 테스트도 작성해보셔도 좋겠습니다. :)
단위 테스트와 인수 테스트를 모두 작성하며 각각의 장단점에 대해서도 생각해보실 수 있을 것 같아요. 이를 바탕으로 추후 업무를 하실 때 어떤 테스트를 어떻게 사용하는 것이 좋을지도 생각해보실 수 있을 것 같습니다!

seokhongkim reviewed (approved):
승기님, 4단계 미션 코멘트 잘 반영해주셨습니다.
4단계 PR도 머지하도록 하겠습니다.

ATDD 학습과 미션을 통해 인수 테스트를 먼저 작성하고, 이후 개발을 진행하는 연습을 충분히 하신 것 같습니다. 마지막 코멘트에 남긴 것처럼 단위 테스트와 인수 테스트, 통합 테스트를 작성해보시며 각 테스트의 장단점, 어떤 테스트 전략을 가져갈지도 고민해보시면 더 도움이 될 것이라고 생각합니다.

남은 학습 기간에도 지금처럼 열심히 학습하셔서 목표로 하시는 성장과 성취 모두 이루시길 응원하겠습니다. 💪 
ATDD 미션 잘 진행해주셨고, 고생 많으셨습니다! 👍
